File: eval_harness/utils/telemetry/tracing/cc_transcript.py
#!/usr/bin/env python3
"""
Claude Code transcript -> normalized agent trace (offline, no OTel, no live hook).

Claude Code writes a per-session JSONL transcript under
~/.claude/projects/<encoded-cwd>/<session-uuid>.jsonl. Each line is a record:

  {"type":"user","message":{"role":"user","content":"..." | [blocks]}, ...}
  {"type":"assistant","message":{"role":"assistant","model":"claude-...",
      "content":[{"type":"text","text":...},
                 {"type":"thinking","thinking":...},
                 {"type":"tool_use","id":...,"name":...,"input":{...}}],
      "usage":{"input_tokens":...,"output_tokens":...,"cache_read_input_tokens":...}}}
  {"type":"user","message":{"content":[{"type":"tool_result","tool_use_id":...,
                                        "content":...,"is_error":...}]}}

We reconstruct it into the SAME shape codex_rollout.py emits (turns -> steps ->
tool_calls), so eval_harness/utils/telemetry/tracing/bt_emit.py and downstream analysis treat both
agents identically. This is the trace-claude-code approach, but run at COLLECTION
TIME (no live plugin/daemon -> nothing left running to slow the machine).

A "turn" begins at a real user prompt (a user message whose content is human
text, not a tool_result) and runs through the assistant model steps until the
next real user prompt. tool_result blocks are matched back to their tool_use by
id to fill in tool outputs.

Test the parser standalone:  python cc_transcript.py --parse-only <transcript.jsonl>
"""

import argparse
import glob
import json
import os
import pathlib
import sys


def _blocks(content):
    """Normalize a message `content` field to a list of blocks."""
    if isinstance(content, str):
        return [{"type": "text", "text": content}]
    if isinstance(content, list):
        return content
    return []


def _is_real_user(message) -> bool:
    """True if this user message is a human prompt (not a tool_result feedback)."""
    for b in _blocks(message.get("content")):
        if isinstance(b, dict) and b.get("type") == "tool_result":
            return False
    return True


def _user_text(message) -> str:
    parts = []
    for b in _blocks(message.get("content")):
        if isinstance(b, dict) and b.get("type") == "text":
            parts.append(b.get("text") or "")
        elif isinstance(b, str):
            parts.append(b)
    return "".join(parts)


def _usage(u):
    if not isinstance(u, dict):
        return {}
    out = {}
    it, ot = u.get("input_tokens"), u.get("output_tokens")
    if it is not None:
        out["prompt_tokens"] = it
    if ot is not None:
        out["completion_tokens"] = ot
    if it is not None and ot is not None:
        out["tokens"] = it + ot
    if u.get("cache_read_input_tokens") is not None:
        out["cache_read_input_tokens"] = u["cache_read_input_tokens"]
    if u.get("cache_creation_input_tokens") is not None:
        out["cache_creation_input_tokens"] = u["cache_creation_input_tokens"]
    return out


def parse_transcript(path):
    records = []
    with open(path) as f:
        for ln in f:
            ln = ln.strip()
            if not ln:
                continue
            try:
                records.append(json.loads(ln))
            except Exception:
                continue

    session_meta = {}
    turns = []
    cur = None
    tools_by_id = {}
    turn_index = 0

    def finish_turn():
        nonlocal cur, tools_by_id
        if cur is None:
            return
        if not cur.get("final_output"):
            for st in reversed(cur["steps"]):
                if st.get("text"):
                    cur["final_output"] = st["text"]
                    break
        turns.append(cur)
        cur, tools_by_id = None, {}

    for rec in records:
        typ = rec.get("type")
        if typ in ("user", "assistant") and not session_meta:
            session_meta = {"id": rec.get("sessionId"), "cwd": rec.get("cwd"),
                            "cli_version": rec.get("version"),
                            "git_branch": rec.get("gitBranch")}
        msg = rec.get("message") or {}

        if typ == "user":
            if _is_real_user(msg):
                # Start of a new turn.
                finish_turn()
                turn_index += 1
                cur = {"turn_id": rec.get("uuid"), "turn_index": turn_index,
                       "user_input": _user_text(msg), "steps": []}
                tools_by_id = {}
            else:
                # tool_result(s) feeding back to the prior assistant step.
                if cur is None:
                    continue
                for b in _blocks(msg.get("content")):
                    if isinstance(b, dict) and b.get("type") == "tool_result":
                        tc = tools_by_id.get(b.get("tool_use_id"))
                        if tc is not None:
                            content = b.get("content")
                            tc["output"] = content if isinstance(content, str) else json.dumps(content, default=str)
                            if b.get("is_error"):
                                tc["error"] = tc["output"]

        elif typ == "assistant":
            if cur is None:
                # Assistant text with no preceding human prompt (rare) — open a turn.
                turn_index += 1
                cur = {"turn_id": rec.get("uuid"), "turn_index": turn_index,
                       "user_input": None, "steps": []}
                tools_by_id = {}
            step = {"text": "", "reasoning": "", "tool_calls": [], "usage": _usage(msg.get("usage"))}
            if msg.get("model"):
                cur["model"] = msg["model"]
            for b in _blocks(msg.get("content")):
                if not isinstance(b, dict):
                    continue
                bt = b.get("type")
                if bt == "text":
                    step["text"] += b.get("text") or ""
                elif bt == "thinking":
                    step["reasoning"] += b.get("thinking") or ""
                elif bt == "tool_use":
                    tc = {"call_id": b.get("id"), "name": b.get("name"), "args": b.get("input")}
                    step["tool_calls"].append(tc)
                    if tc["call_id"]:
                        tools_by_id[tc["call_id"]] = tc
            cur["steps"].append(step)
            # Accumulate turn usage.
            tu = cur.setdefault("total_usage", {})
            for k, v in step["usage"].items():
                if isinstance(v, (int, float)):
                    tu[k] = tu.get(k, 0) + v

    finish_turn()
    return turns, session_meta


# ---------- collection-time reconstruction ----------

def find_transcripts(projects_dir, since_mtime=0.0, before_mtime=None):
    paths = []
    for p in glob.glob(os.path.join(projects_dir, "**", "*.jsonl"), recursive=True):
        try:
            mtime = os.path.getmtime(p)
            if mtime >= since_mtime and (before_mtime is None or mtime < before_mtime):
                paths.append(p)
        except OSError:
            continue
    return sorted(paths, key=lambda p: os.path.getmtime(p))


def reconstruct(
    projects_dir,
    out_path,
    since_mtime=0.0,
    before_mtime=None,
    braintrust=False,
    run_id=None,
    source="claude-code",
    session_name=None,
):
    transcripts = find_transcripts(projects_dir, since_mtime, before_mtime)
    sessions = []
    for tp in transcripts:
        try:
            turns, meta = parse_transcript(tp)
        except Exception as e:
            print(f"[cc_transcript] skip {tp}: {e}")
            continue
        if not turns:
            continue
        sessions.append({"transcript": tp, "session_meta": meta, "turns": turns})
        if braintrust:
            try:
                sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent))
                import bt_emit
                bt_emit.log_session(
                    turns,
                    meta,
                    run_id=run_id,
                    source=source,
                    name=session_name,
                )
            except Exception as e:
                print(f"[cc_transcript] braintrust push failed: {e}")
    payload = {
        "agent": "claude-code",
        "run_id": run_id,
        "source": source,
        "session_name": session_name,
        "n_sessions": len(sessions),
        "projects_dir": projects_dir,
        "since_mtime": since_mtime,
        "before_mtime": before_mtime,
        "sessions": sessions,
    }
    pathlib.Path(out_path).parent.mkdir(parents=True, exist_ok=True)
    with open(out_path, "w") as f:
        json.dump(payload, f, indent=2, default=str)
    print(f"[cc_transcript] {len(sessions)} session(s) -> {out_path}")
    return payload


def main():
    if len(sys.argv) >= 3 and sys.argv[1] == "--parse-only":
        turns, meta = parse_transcript(sys.argv[2])
        print(json.dumps({"session_meta": meta, "turns": turns}, indent=2, default=str))
        return
    ap = argparse.ArgumentParser(description="Reconstruct Claude Code agent traces from transcripts.")
    ap.add_argument("--projects-dir", default=os.path.expanduser("~/.claude/projects"))
    ap.add_argument("--out", required=True)
    ap.add_argument("--since-mtime", type=float, default=0.0,
                    help="epoch seconds; only transcripts modified at/after are included")
    ap.add_argument("--before-mtime", type=float, default=None,
                    help="epoch seconds; only transcripts modified before this time are included")
    ap.add_argument("--braintrust", action="store_true")
    ap.add_argument("--run-id", default=None)
    ap.add_argument("--source", default=os.environ.get("TRACE_SOURCE", "claude-code"))
    ap.add_argument("--session-name", default=os.environ.get("TRACE_SESSION_NAME"))
    args = ap.parse_args()
    reconstruct(
        args.projects_dir,
        args.out,
        args.since_mtime,
        args.before_mtime,
        args.braintrust,
        args.run_id,
        args.source,
        args.session_name,
    )


if __name__ == "__main__":
    main()

File: eval_harness/utils/telemetry/tracing/cc_transcript.ts
#!/usr/bin/env bun

/**
 * Reconstruct Claude Code JSONL transcripts into the harness's normalized
 * turns -> steps -> tool_calls trace shape. Human messages delimit turns;
 * tool-result blocks attach to tool-use blocks only through their matching ID.
 */

import { readdir, readFile, realpath, stat, writeFile, mkdir } from "node:fs/promises";
import type { Dirent } from "node:fs";
import { homedir } from "node:os";
import { dirname, join, resolve } from "node:path";

export type JsonRecord = Record<string, unknown>;
export type ToolCall = JsonRecord & {
  call_id: unknown;
  name: unknown;
  args: unknown;
};
export type TraceStep = JsonRecord & {
  text: string;
  reasoning: string;
  tool_calls: ToolCall[];
};
export type TraceTurn = JsonRecord & {
  turn_id: unknown;
  turn_index: number;
  user_input: unknown;
  steps: TraceStep[];
};

type ClaudeReconstructionOptions = {
  projectsDir: string;
  outPath: string;
  sinceMtime?: number;
  beforeMtime?: number;
  braintrust?: boolean;
  runId?: string | null;
  source?: string;
  sessionName?: string | null;
};

function asRecord(value: unknown): JsonRecord {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new TypeError("expected a JSON object record");
  }
  return value as JsonRecord;
}

function optionalRecord(value: unknown): JsonRecord {
  return typeof value === "object" && value !== null && !Array.isArray(value)
    ? (value as JsonRecord)
    : {};
}

function pythonTruthy(value: unknown): boolean {
  if (
    value === null ||
    value === undefined ||
    value === false ||
    value === 0 ||
    value === 0n ||
    value === ""
  ) {
    return false;
  }
  if (Array.isArray(value)) return value.length > 0;
  if (typeof value === "object") return Object.keys(value).length > 0;
  return true;
}

function recordOrPythonFallback(value: unknown, field: string): JsonRecord {
  if (!pythonTruthy(value)) return {};
  if (typeof value === "object" && value !== null && !Array.isArray(value)) {
    return value as JsonRecord;
  }
  throw new TypeError(`${field} must be an object when present`);
}

function blocks(content: unknown): unknown[] {
  if (typeof content === "string") {
    return [{ type: "text", text: content }];
  }
  return Array.isArray(content) ? content : [];
}

function isRealUser(message: JsonRecord): boolean {
  return !blocks(message.content).some((block) => {
    const record = optionalRecord(block);
    return record.type === "tool_result";
  });
}

function userText(message: JsonRecord): string {
  const parts: string[] = [];
  for (const block of blocks(message.content)) {
    if (typeof block === "string") {
      parts.push(block);
      continue;
    }
    const record = optionalRecord(block);
    if (record.type === "text") {
      parts.push(typeof record.text === "string" ? record.text : "");
    }
  }
  return parts.join("");
}

function addTokenValues(left: unknown, right: unknown): unknown {
  if (typeof left === "number" && typeof right === "number") {
    return left + right;
  }
  if (typeof left === "string" && typeof right === "string") {
    return left + right;
  }
  if (typeof left === "bigint" && typeof right === "bigint") {
    return left + right;
  }
  if (typeof left === "bigint" && typeof right === "number") {
    return Number(left) + right;
  }
  if (typeof left === "number" && typeof right === "bigint") {
    return left + Number(right);
  }
  throw new TypeError("token values cannot be added");
}

function usage(value: unknown): JsonRecord {
  const source = optionalRecord(value);
  const result: JsonRecord = {};
  const inputTokens = source.input_tokens;
  const outputTokens = source.output_tokens;
  if (inputTokens !== null && inputTokens !== undefined) {
    result.prompt_tokens = inputTokens;
  }
  if (outputTokens !== null && outputTokens !== undefined) {
    result.completion_tokens = outputTokens;
  }
  if (
    inputTokens !== null &&
    inputTokens !== undefined &&
    outputTokens !== null &&
    outputTokens !== undefined
  ) {
    result.tokens = addTokenValues(inputTokens, outputTokens);
  }
  if (source.cache_read_input_tokens !== null && source.cache_read_input_tokens !== undefined) {
    result.cache_read_input_tokens = source.cache_read_input_tokens;
  }
  if (
    source.cache_creation_input_tokens !== null &&
    source.cache_creation_input_tokens !== undefined
  ) {
    result.cache_creation_input_tokens = source.cache_creation_input_tokens;
  }
  return result;
}

function pythonString(value: string): string {
  let result = '"';
  for (const character of value) {
    const codePoint = character.codePointAt(0);
    if (codePoint === undefined) {
      continue;
    }
    if (character === '"') result += '\\"';
    else if (character === "\\") result += "\\\\";
    else if (character === "\b") result += "\\b";
    else if (character === "\f") result += "\\f";
    else if (character === "\n") result += "\\n";
    else if (character === "\r") result += "\\r";
    else if (character === "\t") result += "\\t";
    else if (codePoint >= 0x20 && codePoint <= 0x7e) result += character;
    else if (codePoint <= 0xffff) result += `\\u${codePoint.toString(16).padStart(4, "0")}`;
    else {
      const adjusted = codePoint - 0x10000;
      const high = 0xd800 + (adjusted >> 10);
      const low = 0xdc00 + (adjusted & 0x3ff);
      result += `\\u${high.toString(16)}\\u${low.toString(16)}`;
    }
  }
  return `${result}"`;
}

const NUMBER_SOURCES = new WeakMap<object, Map<string, string>>();

function parseJsonWithNumberSources(text: string): unknown {
  const parse = JSON.parse as unknown as (
    source: string,
    reviver: (
      this: unknown,
      key: string,
      value: unknown,
      context: { source?: string },
    ) => unknown,
  ) => unknown;
  return parse(text, function rememberNumberSource(key, value, context) {
    if (
      typeof value === "number" &&
      typeof this === "object" &&
      this !== null &&
      context.source !== undefined
    ) {
      if (/^-?[0-9]+$/.test(context.source) && !Number.isSafeInteger(value)) {
        return BigInt(context.source);
      }
      const holder = this as object;
      const sources = NUMBER_SOURCES.get(holder) ?? new Map<string, string>();
      sources.set(key, context.source);
      NUMBER_SOURCES.set(holder, sources);
    }
    return value;
  });
}

export function parsePythonJson(text: string): unknown {
  return parseJsonWithNumberSources(text);
}

function formatPythonFloat(value: number): string {
  if (Number.isNaN(value)) return "NaN";
  if (value === Infinity) return "Infinity";
  if (value === -Infinity) return "-Infinity";
  if (Object.is(value, -0)) return "-0.0";
  const absolute = Math.abs(value);
  if (absolute !== 0 && (absolute < 1e-4 || absolute >= 1e16)) {
    const [mantissa = "0", rawExponent = "+0"] = value.toExponential().split("e");
    const exponent = Number(rawExponent);
    const sign = exponent < 0 ? "-" : "+";
    return `${mantissa}e${sign}${String(Math.abs(exponent)).padStart(2, "0")}`;
  }
  const decimal = String(value);
  return Number.isInteger(value) ? `${decimal}.0` : decimal;
}

function pythonJsonDumps(
  value: unknown,
  holder?: object,
  key?: string,
  indent?: number,
  level = 0,
): string {
  if (value === null) return "null";
  if (value === true) return "true";
  if (value === false) return "false";
  if (typeof value === "string") return pythonString(value);
  if (typeof value === "number") {
    const source = holder === undefined || key === undefined ? undefined : NUMBER_SOURCES.get(holder)?.get(key);
    if (source !== undefined && /[.eE]/.test(source)) return formatPythonFloat(value);
    return String(value);
  }
  if (typeof value === "bigint") return String(value);
  if (Array.isArray(value)) {
    if (value.length === 0) return "[]";
    if (indent === undefined) {
      return `[${value
        .map((item, index) => pythonJsonDumps(item, value, String(index)))
        .join(", ")}]`;
    }
    const itemIndent = " ".repeat((level + 1) * indent);
    const closingIndent = " ".repeat(level * indent);
    const items = value.map(
      (item, index) =>
        `${itemIndent}${pythonJsonDumps(item, value, String(index), indent, level + 1)}`,
    );
    return `[\n${items.join(",\n")}\n${closingIndent}]`;
  }
  if (typeof value === "object" && value !== null) {
    const entries = Object.entries(value).filter(([, item]) => item !== undefined);
    if (entries.length === 0) return "{}";
    if (indent === undefined) {
      return `{${entries
        .map(
          ([itemKey, item]) =>
            `${pythonString(itemKey)}: ${pythonJsonDumps(item, value, itemKey)}`,
        )
        .join(", ")}}`;
    }
    const itemIndent = " ".repeat((level + 1) * indent);
    const closingIndent = " ".repeat(level * indent);
    const items = entries.map(
      ([itemKey, item]) =>
        `${itemIndent}${pythonString(itemKey)}: ${pythonJsonDumps(
          item,
          value,
          itemKey,
          indent,
          level + 1,
        )}`,
    );
    return `{\n${items.join(",\n")}\n${closingIndent}}`;
  }
  return pythonString(String(value));
}

export function stringifyPythonJson(value: unknown, indent?: number): string {
  return pythonJsonDumps(value, undefined, undefined, indent);
}

async function readJsonl(path: string): Promise<JsonRecord[]> {
  const records: JsonRecord[] = [];
  for (const rawLine of (await readFile(path, "utf8")).split("\n")) {
    const line = rawLine.trim();
    if (line === "") continue;
    try {
      records.push(asRecord(parseJsonWithNumberSources(line)));
    } catch (error) {
      if (error instanceof SyntaxError) continue;
      throw error;
    }
  }
  return records;
}

export async function parseTranscript(path: string): Promise<[TraceTurn[], JsonRecord]> {
  const records = await readJsonl(path);
  let sessionMeta: JsonRecord = {};
  const turns: TraceTurn[] = [];
  let current: TraceTurn | undefined;
  let toolsById = new Map<unknown, ToolCall>();
  let turnIndex = 0;

  const finishTurn = (): void => {
    if (current === undefined) return;
    if (!current.final_output) {
      for (const step of [...current.steps].reverse()) {
        if (step.text) {
          current.final_output = step.text;
          break;
        }
      }
    }
    turns.push(current);
    current = undefined;
    toolsById = new Map();
  };

  for (const record of records) {
    const type = record.type;
    if ((type === "user" || type === "assistant") && Object.keys(sessionMeta).length === 0) {
      sessionMeta = {
        id: record.sessionId ?? null,
        cwd: record.cwd ?? null,
        cli_version: record.version ?? null,
        git_branch: record.gitBranch ?? null,
      };
    }
    const message = recordOrPythonFallback(record.message, "message");

    if (type === "user") {
      if (isRealUser(message)) {
        finishTurn();
        turnIndex += 1;
        current = {
          turn_id: record.uuid ?? null,
          turn_index: turnIndex,
          user_input: userText(message),
          steps: [],
        };
        toolsById = new Map();
      } else if (current !== undefined) {
        for (const block of blocks(message.content)) {
          const result = optionalRecord(block);
          if (result.type !== "tool_result") continue;
          const toolCall = toolsById.get(result.tool_use_id);
          if (toolCall === undefined) continue;
          const content = result.content;
          toolCall.output =
            typeof content === "string" ? content : pythonJsonDumps(content, result, "content");
          if (result.is_error) toolCall.error = toolCall.output;
        }
      }
      continue;
    }

    if (type !== "assistant") continue;
    if (current === undefined) {
      turnIndex += 1;
      current = {
        turn_id: record.uuid ?? null,
        turn_index: turnIndex,
        user_input: null,
        steps: [],
      };
      toolsById = new Map();
    }
    const step: TraceStep = {
      text: "",
      reasoning: "",
      tool_calls: [],
      usage: usage(message.usage),
    };
    if (message.model) current.model = message.model;
    for (const block of blocks(message.content)) {
      const item = optionalRecord(block);
      if (item.type === "text") {
        step.text += typeof item.text === "string" ? item.text : "";
      } else if (item.type === "thinking") {
        step.reasoning += typeof item.thinking === "string" ? item.thinking : "";
      } else if (item.type === "tool_use") {
        const toolCall: ToolCall = {
          call_id: item.id ?? null,
          name: item.name ?? null,
          args: item.input ?? null,
        };
        step.tool_calls.push(toolCall);
        if (toolCall.call_id) toolsById.set(toolCall.call_id, toolCall);
      }
    }
    current.steps.push(step);
    const totalUsage = optionalRecord(current.total_usage);
    current.total_usage = totalUsage;
    for (const [key, value] of Object.entries(optionalRecord(step.usage))) {
      if (typeof value === "number" || typeof value === "bigint") {
        const previous = totalUsage[key];
        if (typeof value === "bigint") {
          totalUsage[key] =
            typeof previous === "number"
              ? previous + Number(value)
              : (typeof previous === "bigint" ? previous : 0n) + value;
        } else {
          totalUsage[key] =
            typeof previous === "bigint"
              ? Number(previous) + value
              : (typeof previous === "number" ? previous : 0) + value;
        }
      }
    }
  }

  finishTurn();
  return [turns, sessionMeta];
}

async function walkJsonl(
  directory: string,
  ancestorDirectories: ReadonlySet<string> = new Set(),
): Promise<string[]> {
  const paths: string[] = [];
  let resolvedDirectory: string;
  let entries: Dirent<string>[];
  try {
    resolvedDirectory = await realpath(directory);
    if (ancestorDirectories.has(resolvedDirectory)) return [];
    entries = await readdir(directory, { withFileTypes: true });
  } catch {
    return [];
  }
  const nextAncestors = new Set(ancestorDirectories);
  nextAncestors.add(resolvedDirectory);
  for (const entry of entries) {
    const path = join(directory, entry.name);
    try {
      const target = await stat(path);
      if (target.isDirectory()) paths.push(...(await walkJsonl(path, nextAncestors)));
      else if (target.isFile() && entry.name.endsWith(".jsonl")) paths.push(path);
    } catch {
      // A broken/inaccessible entry does not discard readable siblings.
    }
  }
  return paths;
}

export async function findTranscripts(
  projectsDir: string,
  sinceMtime = 0,
  beforeMtime?: number,
): Promise<string[]> {
  let candidates: string[];
  try {
    candidates = await walkJsonl(projectsDir);
  } catch {
    return [];
  }
  const selected: Array<{ path: string; mtime: number }> = [];
  for (const path of candidates) {
    try {
      const mtime = (await stat(path)).mtimeMs / 1_000;
      if (mtime >= sinceMtime && (beforeMtime === undefined || mtime < beforeMtime)) {
        selected.push({ path, mtime });
      }
    } catch {
      // A disappearing transcript is ignored just like Python's glob/stat loop.
    }
  }
  selected.sort((left, right) => left.mtime - right.mtime);
  return selected.map(({ path }) => path);
}

const REPO_ROOT = resolve(import.meta.dir, "../../../..");
const BRAINTRUST_BRIDGE = `
import json, sys
from eval_harness.utils.telemetry.tracing import bt_emit
payload = json.load(sys.stdin)
bt_emit.log_session(payload["turns"], payload["session_meta"], run_id=payload.get("run_id"), source=payload.get("source", "agent"), name=payload.get("name"))
`;

export async function emitBraintrustSession(
  turns: TraceTurn[],
  metadata: JsonRecord,
  runId: string | null,
  source: string,
  name: string | null,
  caller = "cc_transcript",
): Promise<void> {
  const searchPath = process.env.PATH;
  const searchOptions = searchPath === undefined ? undefined : { PATH: searchPath };
  const uv = Bun.which("uv", searchOptions);
  const python =
    Bun.which("python3", searchOptions) ?? Bun.which("python", searchOptions);
  const command = uv
    ? [uv, "run", "python", "-c", BRAINTRUST_BRIDGE]
    : python
      ? [python, "-c", BRAINTRUST_BRIDGE]
      : undefined;
  if (command === undefined) {
    console.log(`[${caller}] braintrust push failed: no Python interpreter found`);
    return;
  }
  try {
    const child = Bun.spawn(command, {
      cwd: REPO_ROOT,
      stdin: "pipe",
      stdout: "pipe",
      stderr: "pipe",
      env: process.env,
    });
    child.stdin.write(
      stringifyPythonJson({ turns, session_meta: metadata, run_id: runId, source, name }),
    );
    child.stdin.end();
    const [exitCode, stdout, stderr] = await Promise.all([
      child.exited,
      new Response(child.stdout).text(),
      new Response(child.stderr).text(),
    ]);
    if (stdout) process.stdout.write(stdout);
    if (exitCode !== 0) {
      console.log(`[${caller}] braintrust push failed: ${stderr.trim() || `exit ${exitCode}`}`);
    }
  } catch (error) {
    console.log(`[${caller}] braintrust push failed: ${String(error)}`);
  }
}

export async function reconstructClaudeTranscripts(
  options: ClaudeReconstructionOptions,
): Promise<JsonRecord & { sessions: JsonRecord[] }> {
  const sinceMtime = options.sinceMtime ?? 0;
  const source = options.source ?? "claude-code";
  const runId = options.runId ?? null;
  const sessionName = options.sessionName ?? null;
  const transcripts = await findTranscripts(options.projectsDir, sinceMtime, options.beforeMtime);
  const sessions: JsonRecord[] = [];
  for (const transcript of transcripts) {
    try {
      const [turns, metadata] = await parseTranscript(transcript);
      if (turns.length === 0) continue;
      sessions.push({ transcript, session_meta: metadata, turns });
      if (options.braintrust) {
        await emitBraintrustSession(turns, metadata, runId, source, sessionName);
      }
    } catch (error) {
      console.log(`[cc_transcript] skip ${transcript}: ${String(error)}`);
    }
  }
  const payload = {
    agent: "claude-code",
    run_id: runId,
    source,
    session_name: sessionName,
    n_sessions: sessions.length,
    projects_dir: options.projectsDir,
    since_mtime: sinceMtime,
    before_mtime: options.beforeMtime ?? null,
    sessions,
  };
  await mkdir(dirname(options.outPath), { recursive: true });
  await writeFile(options.outPath, stringifyPythonJson(payload, 2));
  console.log(`[cc_transcript] ${sessions.length} session(s) -> ${options.outPath}`);
  return payload;
}

type ParsedArguments = {
  projectsDir: string;
  outPath?: string;
  sinceMtime: number;
  beforeMtime?: number;
  braintrust: boolean;
  runId: string | null;
  source: string;
  sessionName: string | null;
};

const DIGIT_PART = String.raw`[0-9](?:_?[0-9])*`;
const FINITE_FLOAT_PATTERN = new RegExp(
  String.raw`^[+-]?(?:(?:${DIGIT_PART}(?:\.(?:${DIGIT_PART})?)?)|(?:\.${DIGIT_PART}))(?:[eE][+-]?${DIGIT_PART})?$`,
);
const NON_FINITE_FLOAT_PATTERN = /^[+-]?(?:inf(?:inity)?|nan)$/i;

export function parsePythonFloat(text: string): number | undefined {
  const stripped = text.trim();
  if (stripped === "") return undefined;
  if (NON_FINITE_FLOAT_PATTERN.test(stripped)) {
    const unsigned = stripped.replace(/^[+-]/, "").toLowerCase();
    if (unsigned === "nan") return Number.NaN;
    return stripped.startsWith("-") ? Number.NEGATIVE_INFINITY : Number.POSITIVE_INFINITY;
  }
  if (!FINITE_FLOAT_PATTERN.test(stripped)) return undefined;
  return Number(stripped.replaceAll("_", ""));
}

const CC_HELP = `usage: cc_transcript.py [-h] [--projects-dir PROJECTS_DIR] --out OUT
                        [--since-mtime SINCE_MTIME]
                        [--before-mtime BEFORE_MTIME] [--braintrust]
                        [--run-id RUN_ID] [--source SOURCE]
                        [--session-name SESSION_NAME]

Reconstruct Claude Code agent traces from transcripts.

options:
  -h, --help            show this help message and exit
  --projects-dir PROJECTS_DIR
  --out OUT
  --since-mtime SINCE_MTIME
                        epoch seconds; only transcripts modified at/after are
                        included
  --before-mtime BEFORE_MTIME
                        epoch seconds; only transcripts modified before this
                        time are included
  --braintrust
  --run-id RUN_ID
  --source SOURCE
  --session-name SESSION_NAME`;

const CC_OPTIONS = [
  "--help",
  "--projects-dir",
  "--out",
  "--since-mtime",
  "--before-mtime",
  "--braintrust",
  "--run-id",
  "--source",
  "--session-name",
] as const;

function resolveOption(name: string): (typeof CC_OPTIONS)[number] {
  const candidates = CC_OPTIONS.filter((option) => option === name || option.startsWith(name));
  if (candidates.length !== 1) {
    throw new Error(
      candidates.length === 0
        ? `unrecognized argument: ${name}`
        : `ambiguous option: ${name} could match ${candidates.join(", ")}`,
    );
  }
  return candidates[0] as (typeof CC_OPTIONS)[number];
}

function parseArguments(args: string[]): ParsedArguments | "help" {
  const parsed: ParsedArguments = {
    projectsDir: join(homedir(), ".claude", "projects"),
    sinceMtime: 0,
    braintrust: false,
    runId: null,
    source: process.env.TRACE_SOURCE ?? "claude-code",
    sessionName: process.env.TRACE_SESSION_NAME ?? null,
  };
  for (let index = 0; index < args.length; index += 1) {
    const argument = args[index];
    if (argument === "-h") return "help";
    if (argument === undefined || !argument.startsWith("--")) {
      throw new Error(`unrecognized argument: ${argument ?? ""}`);
    }
    const equalsIndex = argument.indexOf("=");
    const optionText = equalsIndex === -1 ? argument : argument.slice(0, equalsIndex);
    const attachedValue = equalsIndex === -1 ? undefined : argument.slice(equalsIndex + 1);
    const option = resolveOption(optionText);
    if (option === "--help") return "help";
    if (option === "--braintrust") {
      if (attachedValue !== undefined) throw new Error("argument --braintrust: ignored explicit argument");
      parsed.braintrust = true;
      continue;
    }
    const value = attachedValue ?? args[index + 1];
    if (value === undefined) {
      throw new Error(`argument ${option}: expected one argument`);
    }
    if (option === "--projects-dir") parsed.projectsDir = value;
    else if (option === "--out") parsed.outPath = value;
    else if (option === "--since-mtime" || option === "--before-mtime") {
      const numeric = parsePythonFloat(value);
      if (numeric === undefined) throw new Error(`argument ${option}: invalid float value: '${value}'`);
      if (option === "--since-mtime") parsed.sinceMtime = numeric;
      else parsed.beforeMtime = numeric;
    } else if (option === "--run-id") parsed.runId = value;
    else if (option === "--source") parsed.source = value;
    else if (option === "--session-name") parsed.sessionName = value;
    if (attachedValue === undefined) index += 1;
  }
  return parsed;
}

export async function main(args: string[] = process.argv.slice(2)): Promise<number> {
  if (args[0] === "--parse-only" && args[1] !== undefined) {
    const [turns, metadata] = await parseTranscript(args[1]);
    console.log(stringifyPythonJson({ session_meta: metadata, turns }, 2));
    return 0;
  }
  try {
    const parsed = parseArguments(args);
    if (parsed === "help") {
      console.log(CC_HELP);
      return 0;
    }
    if (parsed.outPath === undefined) throw new Error("the following arguments are required: --out");
    await reconstructClaudeTranscripts(parsed as ClaudeReconstructionOptions & { outPath: string });
    return 0;
  } catch (error) {
    console.error(`cc_transcript.ts: error: ${error instanceof Error ? error.message : String(error)}`);
    return 2;
  }
}

if (import.meta.main) {
  process.exit(await main());
}
